addon/v1beta1/types_managedclusteraddon.go

```diff
@@ -0,0 +1,279 @@
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Available",type=string,JSONPath=`.status.conditions[?(@.type=="Available")].status`
+// +kubebuilder:printcolumn:name="Degraded",type=string,JSONPath=`.status.conditions[?(@.type=="Degraded")].status`
+// +kubebuilder:printcolumn:name="Progressing",type=string,JSONPath=`.status.conditions[?(@.type=="Progressing")].status`
+
+// ManagedClusterAddOn is the Custom Resource object which holds the current state
+// of an add-on. This object is used by add-on operators to convey their state.
+// This resource should be created in the ManagedCluster namespace.
+type ManagedClusterAddOn struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// spec holds configuration that could apply to any operator.
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ManagedClusterAddOnSpec `json:"spec"`
+
+	// status holds the information about the state of an operator.  It is consistent with status information across
+	// the Kubernetes ecosystem.
+	// +optional
+	Status ManagedClusterAddOnStatus `json:"status"`
+}
+
+// ManagedClusterAddOnSpec defines the install configuration of
+// an addon agent on managed cluster.
+type ManagedClusterAddOnSpec struct {
+	// configs is a list of add-on configurations.
+	// In scenario where the current add-on has its own configurations.
+	// An empty list means there are no defautl configurations for add-on.
+	// The default is an empty list
+	// +optional
+	Configs []AddOnConfig `json:"configs,omitempty"`
+}
+
+// RegistrationConfig defines the configuration of the addon agent to register to hub. The Klusterlet agent will
+// create a csr for the addon agent with the registrationConfig.
+type RegistrationConfig struct {
+	// signerName is the name of signer that addon agent will use to create csr.
+	// +required
+	// +kubebuilder:validation:MaxLength=571
+	// +kubebuilder:validation:MinLength=5
+	// +kubebuilder:validation:Required
+	SignerName string `json:"signerName"`
+
+	// subject is the user subject of the addon agent to be registered to the hub.
+	// If it is not set, the addon agent will have the default subject
+	// "subject": {
+	//	"user": "system:open-cluster-management:addon:{addonName}:{clusterName}:{agentName}",
+	//	"groups: ["system:open-cluster-management:addon", "system:open-cluster-management:addon:{addonName}", "system:authenticated"]
+	// }
+	//
+	// +optional
+	Subject Subject `json:"subject,omitempty"`
+
+	// RegistrationNamespace is the namespace on the managed cluster to put the registration secret for the addon agent.
+	// If it is not set, open-cluster-management-agent-addon namespace is used to install the addon agent.
+	// +optional
+	// +kubebuilder:default=open-cluster-management-agent-addon
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	RegistrationNamespace string `json:"registrationNamespace,omitempty"`
```

Comment:
Not sure if using Namespace directly is better since it is the RegistrationConfig struct.